File: js/src/json-schema-dsl/schemaGetter.js
const common = require('./schemaCommon');

function getObjectField(schemaObj, obj, name) {
    const field = common.findField(schemaObj, name);
    if (field === null) {
        throw new Error('Unrecognized field: ' + name);
    }
    if (obj[name] === undefined || obj[name] === null) {
        throw obj[name];
    }
    return [field.fieldType, obj[name]];
}

function getArrayField(schemaObj, obj, name) {
    const index = common.convertToInteger(name);
    if (index === null) {
        throw new Error('Index should be integer');
    }
    return [schemaObj.elementType, obj[index]];
}

function getStringMapField(schemaObj, obj, name) {
    return [{type: 'String'}, obj[name]]
}

function getBasicField(schemaObj, obj, name) {
    throw new Error('Path is too long')
}

const FieldTypeGetterMap = {
    Object: getObjectField,
    Array: getArrayField,
    StringMap: getStringMapField,
    String: getBasicField,
    Integer: getBasicField,
    Boolean: getBasicField,
    Number: getBasicField
};

function getField(schemaObj, obj, name) {
    return FieldTypeGetterMap[schemaObj.type](schemaObj, obj, name);
}

function getter(schemaObj, path, obj) {
    try {
        for (const name of path) {
            [schemaObj, obj] = getField(schemaObj, obj, name);
        }
        return obj;
    } catch (err) {
        if (err === null || err === undefined) {
            return err;
        }
        throw err;
    }
}

module.exports = getter;

File: python/src/schema_getter.py
class NoneValue(Exception):
    pass


def _find_field(schema_object, field_name):
    for field in schema_object['fields']:
        if field['name'] == field_name:
            return field
    return None


def _get_object_field(schema_object, obj, name):
    field = _find_field(schema_object, name)
    if field is None:
        raise ValueError('Unrecognized field: ' + name)
    if name not in obj or obj[name] is None:
        raise NoneValue()
    return [field['field_type'], obj[name]]


def _get_array_field(schema_object, obj, index):
    if isinstance(index, bool) or not isinstance(index, (int, long)):
        raise TypeError( 'Index should be integer')
    return [schema_object['element_type'], obj[index]]


def _get_string_map_field(schema_object, obj, name):
    return [{'type': 'String'}, obj[name]]


def _get_basic_field(schema_object, obj, name):
    raise ValueError('Path is too long')


FIELD_TYPE_GETTER_MAP = {
    'String': _get_basic_field,
    'Integer': _get_basic_field,
    'Number': _get_basic_field,
    'Boolean': _get_basic_field,
    'StringMap': _get_string_map_field,
    'Object': _get_object_field,
    'Array': _get_array_field
}


def _get_field(schema_object, obj, name):
    return FIELD_TYPE_GETTER_MAP[schema_object['type']](schema_object, obj, name)


def getter(schema_object, path, obj):
    try:
        for name in path:
            schema_object, obj = _get_field(schema_object, obj, name)
        return obj
    except NoneValue:
        return None
